File: andriod/app/src/main/java/com/Bands70k/ScheduleQRGuideLink.java
package com.Bands70k;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.util.Log;

import java.nio.charset.StandardCharsets;
import java.util.Locale;

/**
 * Guide QR deep link: system Camera app opens this app to the schedule QR scanner.
 * URL string comes from festival.json scheduleQRGuideURL (e.g. bands70k://schedule-scan).
 */
public final class ScheduleQRGuideLink {

    private static final String TAG = "ScheduleQRGuide";

    private static boolean pendingOpenScanner = false;

    private ScheduleQRGuideLink() {}

    /** Configured guide URL when schedule QR share is enabled; null otherwise. */
    public static String getConfiguredGuideURLString() {
        FestivalConfig cfg = FestivalConfig.getInstance();
        if (!cfg.scheduleQRShareEnabled) return null;
        String url = cfg.scheduleQRGuideURL;
        if (url == null || url.trim().isEmpty()) return null;
        return url.trim();
    }

    /** Returns true when {@code uri} is the configured guide link (sets pending open). */
    public static boolean handleIncomingUri(Uri uri) {
        if (!matchesGuideUri(uri)) return false;
        Log.d(TAG, "[QRGuide] Incoming guide URL: " + uri);
        pendingOpenScanner = true;
        return true;
    }

    /** Call when main UI is ready — opens scanner after a guide URL at cold launch. */
    public static void deliverPendingOpenScannerIfNeeded(Context context) {
        if (!pendingOpenScanner || context == null) return;
        pendingOpenScanner = false;
        if (!FestivalConfig.getInstance().scheduleQRShareEnabled) return;
        context.startActivity(new Intent(context, ScheduleQRScanActivity.class));
    }

    public static boolean matchesGuideUri(Uri uri) {
        if (uri == null) return false;
        String expected = getConfiguredGuideURLString();
        if (expected == null) return false;
        Uri expectedUri = Uri.parse(expected);
        return urlsEquivalent(uri, expectedUri);
    }

    public static boolean matchesGuidePayload(byte[] data) {
        if (data == null || data.length == 0) return false;
        String text = new String(data, StandardCharsets.UTF_8).trim();
        if (text.isEmpty()) return false;
        return matchesGuideURLString(text);
    }

    /** Scanner only: guide QR is a short UTF-8 URL, not a truncated binary schedule payload. */
    public static boolean matchesGuidePayloadExact(byte[] data) {
        String expected = getConfiguredGuideURLString();
        if (expected == null || data == null || data.length == 0) return false;
        String trimmedExpected = expected.trim();
        if (trimmedExpected.isEmpty()) return false;
        if (data[0] >= 0 && data[0] <= 2) return false;
        String text = new String(data, StandardCharsets.UTF_8).trim();
        if (!text.equals(trimmedExpected)) return false;
        return data.length <= trimmedExpected.getBytes(StandardCharsets.UTF_8).length + 4;
    }

    public static boolean matchesGuideURLString(String string) {
        if (string == null || string.trim().isEmpty()) return false;
        try {
            return matchesGuideUri(Uri.parse(string.trim()));
        } catch (Exception e) {
            return false;
        }
    }

    private static boolean urlsEquivalent(Uri a, Uri b) {
        if (a == null || b == null) return false;
        String schemeA = a.getScheme() != null ? a.getScheme().toLowerCase(Locale.US) : "";
        String schemeB = b.getScheme() != null ? b.getScheme().toLowerCase(Locale.US) : "";
        if (!schemeA.equals(schemeB)) return false;
        String hostA = a.getHost() != null ? a.getHost().toLowerCase(Locale.US) : "";
        String hostB = b.getHost() != null ? b.getHost().toLowerCase(Locale.US) : "";
        if (!hostA.isEmpty() || !hostB.isEmpty()) {
            if (!hostA.equals(hostB)) return false;
        }
        return normalizedPath(a.getPath()).equals(normalizedPath(b.getPath()));
    }

    private static String normalizedPath(String path) {
        String p = path != null ? path : "";
        if (!p.startsWith("/")) p = "/" + p;
        while (p.length() > 1 && p.endsWith("/")) {
            p = p.substring(0, p.length() - 1);
        }
        return p.toLowerCase(Locale.US);
    }
}

File: swift/70000TonsBands/ScheduleQRGuideLink.swift
//
//  ScheduleQRGuideLink.swift
//  70K Bands
//
//  Guide QR deep link: system Camera app opens this app to the schedule QR scanner.
//  URL string comes from festival.json scheduleQRGuideURL (e.g. bands70k://schedule-scan).
//

import Foundation

enum ScheduleQRGuideLink {

    static let openScannerNotification = Notification.Name("OpenScheduleQRScannerFromGuide")

    private static var pendingOpenScanner = false

    /// Configured guide URL when schedule QR share is enabled; nil otherwise.
    static var configuredGuideURLString: String? {
        guard FestivalConfig.current.scheduleQRShareEnabled else { return nil }
        let trimmed = FestivalConfig.current.scheduleQRGuideURL.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }

    static func handleIncomingURL(_ url: URL) -> Bool {
        guard matchesGuideURL(url) else { return false }
        print("[QRGuide] Incoming guide URL: \(url.absoluteString)")
        pendingOpenScanner = true
        DispatchQueue.main.async {
            deliverPendingOpenScannerIfNeeded()
        }
        return true
    }

    /// Call when main UI (masterView) is ready — delivers a guide URL received at cold launch.
    static func deliverPendingOpenScannerIfNeeded() {
        guard pendingOpenScanner else { return }
        guard masterView != nil else { return }
        pendingOpenScanner = false
        NotificationCenter.default.post(name: openScannerNotification, object: nil)
    }

    static func matchesGuideURL(_ url: URL) -> Bool {
        guard let expected = configuredGuideURLString, let expectedURL = URL(string: expected) else { return false }
        return urlsEquivalent(url, expectedURL)
    }

    static func matchesGuidePayload(_ data: Data) -> Bool {
        guard let text = String(data: data, encoding: .utf8)?
            .trimmingCharacters(in: .whitespacesAndNewlines),
              !text.isEmpty else { return false }
        return matchesGuideURLString(text)
    }

    /// Scanner only: guide QR is a short UTF-8 URL string, not a truncated binary schedule payload.
    static func matchesGuidePayloadExact(_ data: Data) -> Bool {
        guard let expected = configuredGuideURLString else { return false }
        let trimmedExpected = expected.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmedExpected.isEmpty else { return false }
        // Binary schedule QRs start with type 0/1/2 and are much larger when complete.
        if let first = data.first, first <= 2 { return false }
        guard let text = String(data: data, encoding: .utf8)?
            .trimmingCharacters(in: .whitespacesAndNewlines),
              text == trimmedExpected else { return false }
        return data.count <= trimmedExpected.utf8.count + 4
    }

    static func matchesGuideURLString(_ string: String) -> Bool {
        guard let url = URL(string: string) else { return false }
        return matchesGuideURL(url)
    }

    private static func urlsEquivalent(_ a: URL, _ b: URL) -> Bool {
        let schemeA = (a.scheme ?? "").lowercased()
        let schemeB = (b.scheme ?? "").lowercased()
        guard schemeA == schemeB else { return false }
        let hostA = (a.host ?? "").lowercased()
        let hostB = (b.host ?? "").lowercased()
        if !hostA.isEmpty || !hostB.isEmpty, hostA != hostB { return false }
        let pathA = normalizedPath(a.path)
        let pathB = normalizedPath(b.path)
        return pathA == pathB
    }

    private static func normalizedPath(_ path: String) -> String {
        var p = path
        if !p.hasPrefix("/") { p = "/" + p }
        while p.count > 1, p.hasSuffix("/") { p.removeLast() }
        return p.lowercased()
    }
}
